src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1954,6 +1964,456 @@ public Task OpenAsync(CancellationToken token)
+            Message closeResponseMessage;
+
+            public ServerSecurityDuplexSessionChannel(
+                SecuritySessionServerSettings settings,
+                SecurityContextSecurityToken sessionToken,
+                object listenerSecurityState, SecurityListenerSettingsLifetimeManager settingsLifetimeManager, EndpointAddress address)
+                : base(settings,
+                      sessionToken, listenerSecurityState, settingsLifetimeManager, address)
+            {
+                this.session = new SoapSecurityServerDuplexSession(sessionToken, settings, this);
+            }
+
+            public EndpointAddress RemoteAddress
+            {
+                get
+                {
+                    return LocalAddress;
+                }
+            }
+
+            public Uri Via
+            {
+                get
+                {
+                    return this.RemoteAddress.Uri;
+                }
+            }
+
+            public IDuplexSession Session
+            {
+                get
+                {
+                    return this.session;
+                }
+            }
+
+            protected override void AbortCore()
+            {
+                base.AbortCore();
+                this.Settings.RemoveSessionChannel(this.session.Id);
+                CleanupPendingCloseState();
+            }
+
+            void CleanupPendingCloseState()
+            {
+                lock (LocalLock)
+                {
+                    if (this.closeResponseMessage != null)
+                    {
+                        this.closeResponseMessage.Close();
+                        this.closeResponseMessage = null;
+                    }
+                    if (this.closeRequestContext != null)
+                    {
+                        this.closeRequestContext.Abort();
+                        this.closeRequestContext = null;
+                    }
+                }
+            }
+
+            public void Abort()
+            {
+                AbortCore();
+            }
+
+            public Task CloseAsync()
+            {
+                return CloseAsync(new TimeoutHelper(ServiceDefaults.CloseTimeout).GetCancellationToken());
+            }
+
+            public Task CloseAsync(CancellationToken token)
+            {
+                // step 1: close output session
+                this.CloseOutputSession(token);
+
+                // if the channel was aborted while closing the output session, return
+                if (this.State == CommunicationState.Closed)
+                {
+                    return Task.CompletedTask;
+                }
+
+                //Check with Matt about Inp/Oout Session
+
+                // step 2: wait for input session to be closed
+                /*
+                bool wasAborted;
+                bool didInputSessionClose = this.WaitForInputSessionClose(timeoutHelper.RemainingTime(), out wasAborted);
+                if (wasAborted)
```

Comment:
This half close mechanism needs to stay

Reply:
I did some changes.. I am not convinced of changes though.. check the latest, may be we can have quick sync to understand the usage of interruptible...